src/Storages/StorageJoin.cpp

```diff
@@ -174,12 +174,45 @@ HashJoinPtr StorageJoin::getJoinLocked(std::shared_ptr<TableJoin> analyzed_join,
             "Table {} needs the same join_use_nulls setting as present in LEFT or FULL JOIN",
             getStorageID().getNameForLogs());
 
-    /// TODO: check key columns
+    const auto & join_on = analyzed_join->getOnlyClause();
+    if (join_on.on_filter_condition_left || join_on.on_filter_condition_right)
+        throw Exception(ErrorCodes::INCOMPATIBLE_TYPE_OF_JOIN, "ON section of JOIN with filter conditions is not implemented");
+
+    const auto & key_names_right = join_on.key_names_right;
+    const auto & key_names_left = join_on.key_names_left;
+    if (key_names.size() != key_names_right.size() || key_names.size() != key_names_left.size())
+        throw Exception(ErrorCodes::INCOMPATIBLE_TYPE_OF_JOIN,
+            "Number of keys in JOIN ON section ({}) doesn't match number of keys in Join engine ({})",
+            key_names_right.size(), key_names.size());
+
+    /* Resort left keys according to right keys order in StorageJoin
+     * We can't change the order of keys in StorageJoin because the hash table already build.
```

Comment:
not obvious, why it matters. I mean we should operate on tuples anyway. maybe because we will serialise them in this order? imo worth to clarify in comment.